lib/libc/gen/err.3
.\" $NetBSD: err.3,v 1.21 2014/01/16 17:21:38 christos Exp $
.\"
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)err.3	8.1 (Berkeley) 6/9/93
.\"
.Dd January 16, 2014
.Dt ERR 3
.Os
.Sh NAME
.Nm err ,
.Nm verr ,
.Nm errx ,
.Nm verrx ,
.Nm errc ,
.Nm verrc ,
.Nm warn ,
.Nm vwarn ,
.Nm warnx ,
.Nm vwarnx ,
.Nm warnc ,
.Nm vwarnc
.Nd formatted error messages
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In err.h
.Ft void
.Fn err "int status" "const char *fmt" "..."
.Ft void
.Fn verr "int status" "const char *fmt" "va_list args"
.Ft void
.Fn errx "int status" "const char *fmt" "..."
.Ft void
.Fn verrx "int status" "const char *fmt" "va_list args"
.Ft void
.Fn errc "int status" "int code" "const char *fmt" "..."
.Ft void
.Fn verrc "int status" "int code" "const char *fmt" "va_list args"
.Ft void
.Fn warn "const char *fmt" "..."
.Ft void
.Fn vwarn "const char *fmt" "va_list args"
.Ft void
.Fn warnx "const char *fmt" "..."
.Ft void
.Fn vwarnx "const char *fmt" "va_list args"
.Ft void
.Fn warnc "int code" "const char *fmt" "..."
.Ft void
.Fn vwarnc "int code" "const char *fmt" "va_list args"
.Sh DESCRIPTION
The
.Fn err
and
.Fn warn
family of functions display a formatted error message on the standard
error output.
In all cases, the last component of the program name, a colon character,
and a space are output.
If the
.Fa fmt
argument is not
.Dv NULL ,
the formatted error message is output.
In the case of the
.Fn err ,
.Fn verr ,
.Fn warn ,
and
.Fn vwarn
functions, the error message string affiliated with the current value of
the global variable
.Va errno
is output next, preceded by a colon character and a space if
.Fa fmt
is not
.Dv NULL .
In all cases, the output is followed by a newline character.
The
.Fn errc ,
.Fn verrc ,
.Fn warnc ,
and
.Fn vwarnc
functions take an additional
.Ar code
argument to be used as the error number instead of using the global
.Va errno
variable.
The
.Fn errx ,
.Fn verrx ,
.Fn warnx ,
and
.Fn vwarnx
functions will not output this error message string.
.Pp
The
.Fn err ,
.Fn verr ,
.Fn errx ,
and
.Fn verrx
functions do not return, but instead cause the program to terminate
with the status value given by the argument
.Fa status .
It is often appropriate to use the value
.Dv EXIT_FAILURE ,
defined in
.In stdlib.h ,
as the
.Fa status
argument given to these functions.
.Sh EXAMPLES
Display the current
.Va errno
information string and terminate with status indicating failure:
.Bd -literal -offset indent
if ((p = malloc(size)) == NULL)
	err(EXIT_FAILURE, NULL);
if ((fd = open(file_name, O_RDONLY, 0)) == -1)
	err(EXIT_FAILURE, "%s", file_name);
.Ed
.Pp
Display an error message and terminate with status indicating failure:
.Bd -literal -offset indent
if (tm.tm_hour \*[Lt] START_TIME)
	errx(EXIT_FAILURE, "too early, wait until %s",
	    start_time_string);
.Ed
.Pp
Warn of an error:
.Bd -literal -offset indent
if ((fd = open(raw_device, O_RDONLY, 0)) == -1)
	warnx("%s: %s: trying the block device",
	    raw_device, strerror(errno));
if ((fd = open(block_device, O_RDONLY, 0)) == -1)
	warn("%s", block_device);
.Ed
.Sh SEE ALSO
.Xr exit 3 ,
.Xr getprogname 3 ,
.Xr strerror 3
.Sh HISTORY
The
.Fn err
and
.Fn warn
functions first appeared in
.Bx 4.4 .
The
.Fn errc
and
.Fn warnc
functions first appeared in
.Fx 3.0
and
.Nx 7.0 .
.Sh CAVEATS
It is important never to pass a string with user-supplied data as a
format without using
.Ql %s .
An attacker can put format specifiers in the string to mangle your stack,
leading to a possible security hole.
This holds true even if you have built the string
.Dq by hand
using a function like
.Fn snprintf ,
as the resulting string may still contain user-supplied conversion specifiers
for later interpolation by the
.Fn err
and
.Fn warn
functions.
.Pp
Always be sure to use the proper secure idiom:
.Bd -literal -offset indent
err(1, "%s", string);
.Ed


minix/drivers/vmm_guest/vbox/err.c
/* VirtualBox driver - by D.C. van Moolenbroek */
#include <minix/drivers.h>
#include <errno.h>

#include "vmmdev.h"
#include "proto.h"

static int codes[] = {
  OK,			/*     0: success */
  EGENERIC,		/*    -1: general failure */
  EINVAL,		/*    -2: invalid parameter */
  EINVAL,		/*    -3: invalid magic */
  EBADF,		/*    -4: invalid handle */
  ENOLCK,		/*    -5: lock failed */
  EFAULT,		/*    -6: invalid pointer */
  EGENERIC,		/*    -7: patching IDT failed */
  ENOMEM,		/*    -8: memory allocation failed */
  EEXIST,		/*    -9: already loaded */
  EPERM,		/*   -10: permission denied */
  EINVAL,		/*   -11: version mismatch */
  ENOSYS,		/*   -12: function not implemented */
  EGENERIC,		/*   -13 */
  EGENERIC,		/*   -14 */
  EGENERIC,		/*   -15 */
  EGENERIC,		/*   -16 */
  EGENERIC,		/*   -17 */
  EGENERIC,		/*   -18: not equal */
  EINVAL,		/*   -19: not a symlink */
  ENOMEM,		/*   -20: temporary memory allocation failed */
  EINVAL,		/*   -21: invalid file mode */
  EINVAL,		/*   -22: incorrect call order */
  EGENERIC,		/*   -23: no TLS available */
  EGENERIC,		/*   -24: failed to set TLS */
  EGENERIC,		/*   -25 */
  ENOMEM,		/*   -26: contiguous memory allocation failed */
  ENOMEM,		/*   -27: no memory available for page table */
  EGENERIC,		/*   -28 */
  ESRCH,		/*   -29: thread is dead */
  EINVAL,		/*   -30: thread is not waitable */
  EGENERIC,		/*   -31: page table not present */
  EINVAL,		/*   -32: invalid context */
  EBUSY,		/*   -33: timer is busy */
  EGENERIC,		/*   -34: address conflict */
  EGENERIC,		/*   -35: unresolved error */
  ENOTTY,		/*   -36: invalid function */
  EINVAL,		/*   -37: not supported */
  EACCES,		/*   -38: access denied */
  EINTR,		/*   -39: interrupted */
  ETIMEDOUT,		/*   -40: timeout */
  E2BIG,		/*   -41: buffer overflow */
  E2BIG,		/*   -42: too much data */
  EAGAIN,		/*   -43: max thread number reached */
  EAGAIN,		/*   -44: max process number reached */
  EGENERIC,		/*   -45: signal refused */
  EBUSY,		/*   -46: signal already pending */
  EINVAL,		/*   -47: invalid signal */
  EGENERIC,		/*   -48: state changed */
  EINVAL,		/*   -49: invalid UUID format */
  ESRCH,		/*   -50: process not found */
  OK,			/*   -51: waited-for process is still running */
  EAGAIN,		/*   -52: try again */
  EGENERIC,		/*   -53: generic parse error */
  ERANGE,		/*   -54: value out of range */
  EOVERFLOW,		/*   -55: value too big */
  EGENERIC,		/*   -56: no digits in string */
  ERANGE,		/*   -57: minus sign in unsigned value */
  EILSEQ,		/*   -58: character translation failed */
  EGENERIC,		/*   -59: encountered unicode byte order mark */
  EGENERIC,		/*   -60: encountered unicode surrogate */
  EILSEQ,		/*   -61: invalid UTF8 encoding */
  EILSEQ,		/*   -62: invalid UTF16 encoding */
  EGENERIC,		/*   -63: no UTF16 for character */
  ENOMEM,		/*   -64: string memory allocation failed */
  ENOMEM,		/*   -65: UTF16 string memory allocation failed */
  ENOMEM,		/*   -66: code point array allocation failed */
  EBUSY,		/*   -67: cannot free in-use memory */
  EGENERIC,		/*   -68: timer already active */
  EGENERIC,		/*   -69: timer already suspended */
  ECANCELED,		/*   -70: operation cancelled */
  ENOMEM,		/*   -71: failed to initialize memory object */
  ENOMEM,		/*   -72: low physical memory allocation failed */
  ENOMEM,		/*   -73: physical memory allocation failed */
  EGENERIC,		/*   -74: address too big */
  EGENERIC,		/*   -75: memory mapping failed */
  EGENERIC,		/*   -76: trailing characters */
  EGENERIC,		/*   -77: trailing spaces */
  ESRCH,		/*   -78: not found */
  EGENERIC,		/*   -79: invalid state */
  ENOMEM,		/*   -80: out of resources */
  ENFILE,		/*   -81: no more handles */
  EGENERIC,		/*   -82: preemption disabled */
  EGENERIC,		/*   -83: end of string */
  EGENERIC,		/*   -84: page count out of range */
  EGENERIC,		/*   -85: object destroyed */
  EGENERIC,		/*   -86: dangling objects */
  EGENERIC,		/*   -87: invalid Base64 encoding */
  EGENERIC,		/*   -88: return triggered by callback */
  EGENERIC,		/*   -89: authentication failure */
  EGENERIC,		/*   -90: not a power of two */
  EGENERIC,		/*   -91: ignored */
  EGENERIC,		/*   -92: concurrent access not allowed */
  EGENERIC,		/*   -93: invalid reference sharing */
  EGENERIC,		/*   -94 */
  EGENERIC,		/*   -95: no change */
  ENOMEM,		/*   -96: executable memory allocation failed */
  EINVAL,		/*   -97: unsupported alignment */
  EGENERIC,		/*   -98: duplicate */
  EGENERIC,		/*   -99: missing */
  EIO,			/*  -100: I/O error */
  ENXIO,		/*  -101: open failed */
  ENOENT,		/*  -102: file not found */
  ENOTDIR,		/*  -103: path not found (may also mean ENOENT) */
  EINVAL,		/*  -104: invalid name */
  EEXIST,		/*  -105: already exists */
  ENFILE,		/*  -106: too many open files */
  EIO,			/*  -107: seek error */
  EINVAL,		/*  -108: negative seek */
  ESPIPE,		/*  -109: seek on device */
  EGENERIC,		/*  -110: end of file */
  EIO,			/*  -111: generic read error */
  EIO,			/*  -112: generic write error */
  EROFS,		/*  -113: write protected */
  ETXTBSY,		/*  -114: sharing violation */
  ENOLCK,		/*  -115: file lock failed */
  EAGAIN,		/*  -116: file lock violation */
  EIO,			/*  -117: cannot create file */
  EIO,			/*  -118: cannot delete directory */
  EXDEV,		/*  -119: not the same device */
  ENAMETOOLONG,		/*  -120: file name too long */
  ENXIO,		/*  -121: media not present */
  EIO,			/*  -122: media not recognized */
  OK,			/*  -123: nothing to unlocked */
  EGENERIC,		/*  -124: lock lost */
  ENOTEMPTY,		/*  -125: directory not empty */
  ENOTDIR,		/*  -126: not a directory */
  EISDIR,		/*  -127: is a directory */
  EFBIG,		/*  -128: file too big */
  EGENERIC,		/*  -129: no asynchronous I/O request */
  EGENERIC,		/*  -130: asynchronous I/O in progress */
  EGENERIC,		/*  -131: asynchronous I/O completed */
  EGENERIC,		/*  -132: asynchronous I/O busy */
  EGENERIC,		/*  -133: asynchronous I/O limit exceeded */
  EGENERIC,		/*  -134: asynchronous I/O canceled */
  EGENERIC,		/*  -135: asynchronous I/O not submitted */
  EGENERIC,		/*  -136: asynchronous I/O not prepared */
  EGENERIC,		/*  -137: asynchronous I/O out of resources */
  EBUSY,		/*  -138: device or resource busy */
  EGENERIC,		/*  -139: not a file */
  EGENERIC,		/*  -140: is a file */
  EGENERIC,		/*  -141: unexpected file type */
  EGENERIC,		/*  -142: missing path root specification */
  EGENERIC,		/*  -143: path is relative */
  EGENERIC,		/*  -144: path is not relative */
  EGENERIC,		/*  -145 */
  EGENERIC,		/*  -146 */
  EGENERIC,		/*  -147 */
  EGENERIC,		/*  -148 */
  EGENERIC,		/*  -149 */
  EIO,			/*  -150: disk I/O error */
  ENXIO,		/*  -151: invalid drive number */
  ENOSPC,		/*  -152: disk full */
  EIO,			/*  -153: disk changed */
  EGENERIC,		/*  -154: drive locked */
  ENXIO,		/*  -155: invalid disk format */
  ELOOP,		/*  -156: too many symlinks */
  EOPNOTSUPP,		/*  -157: can not set symlink file times */
  EOPNOTSUPP,		/*  -158: can not change symlink owner */
  EGENERIC,		/*  -159 */
  EGENERIC,		/*  -160 */
  EGENERIC,		/*  -161 */
  EGENERIC,		/*  -162 */
  EGENERIC,		/*  -163 */
  EGENERIC,		/*  -164 */
  EGENERIC,		/*  -165 */
  EGENERIC,		/*  -166 */
  EGENERIC,		/*  -167 */
  EGENERIC,		/*  -168 */
  EGENERIC,		/*  -169 */
  EGENERIC,		/*  -170 */
  EGENERIC,		/*  -171 */
  EGENERIC,		/*  -172 */
  EGENERIC,		/*  -173 */
  EGENERIC,		/*  -174 */
  EGENERIC,		/*  -175 */
  EGENERIC,		/*  -176 */
  EGENERIC,		/*  -177 */
  EGENERIC,		/*  -178 */
  EGENERIC,		/*  -179 */
  EGENERIC,		/*  -180 */
  EGENERIC,		/*  -181 */
  EGENERIC,		/*  -182 */
  EGENERIC,		/*  -183 */
  EGENERIC,		/*  -184 */
  EGENERIC,		/*  -185 */
  EGENERIC,		/*  -186 */
  EGENERIC,		/*  -187 */
  EGENERIC,		/*  -188 */
  EGENERIC,		/*  -189 */
  EGENERIC,		/*  -190 */
  EGENERIC,		/*  -191 */
  EGENERIC,		/*  -192 */
  EGENERIC,		/*  -193 */
  EGENERIC,		/*  -194 */
  EGENERIC,		/*  -195 */
  EGENERIC,		/*  -196 */
  EGENERIC,		/*  -197 */
  EGENERIC,		/*  -198 */
  EGENERIC,		/*  -199 */
  EGENERIC,		/*  -200: search error */
  OK,			/*  -201: no more files */
  ENFILE,		/*  -202: no more search handles available */
};

int convert_err(int code)
{
/* Return a POSIX error code for the given VirtualBox error code.
 */
  unsigned int index;

  index = -code;

  if (index < sizeof(codes) / sizeof(codes[0]))
	return codes[index];

  return EGENERIC;
}
